packages/dds/map/src/mapKernel.ts

```diff
@@ -440,18 +444,43 @@ export class MapKernel {
         return false;
     }
 
+    /**
+     * Rollback a local op
+     * @param op - The operation to rollback
+     * @param localOpMetadata - The local metadata associated with the op.
+     */
+    public rollback(op: any, localOpMetadata: unknown) {
+        if (op.type !== "delete" && op.type !== "set") {
+            throw new Error("Unsupported op for rollback");
+        }
+
+        const metadata = localOpMetadata as IMapLocalOpMetadata;
+        if (metadata === undefined) {
+            this.deleteCore(op.key, true);
+        } else {
+            this.setCore(op.key, metadata.previousValue, true, true);
+        }
+
+        // BUGBUG: This isn't right. It should revert it to the value it was before the rolled back op.
+        this.pendingKeys.delete(op.key);
+    }
+
     /**
      * Set implementation used for both locally sourced sets as well as incoming remote sets.
      * @param key - The key being set
      * @param value - The value being set
      * @param local - Whether the message originated from the local client
-     * @param op - The message if from a remote set, or null if from a local set
+     * @returns Previous value of the key
      */
-    private setCore(key: string, value: ILocalValue, local: boolean): void {
+    private setCore(key: string, value: ILocalValue, local: boolean, rollback: boolean = false): any {
         const previousValue = this.get(key);
         this.data.set(key, value);
-        const event: IValueChanged = { key, previousValue };
-        this.eventEmitter.emit("valueChanged", event, local, this.eventEmitter);
+        if (rollback) {
+            this.eventEmitter.emit("rollback", key, this.eventEmitter);
+        } else {
+            this.eventEmitter.emit("valueChanged", { key, previousValue }, local, this.eventEmitter);
```

Comment:
if we think it matters this is a rollback then i think if should just be a property on the object provided by "valueChanged" as ideally most code won't really care. it will just listen to value changed and see the changed values for rollback and all other cases. i don't think all consumers should need specialized handling in the rollback case